// @appland/navie/dist/explain.js
"use strict";
var __await = (this && this.__await) || function (v) { return this instanceof __await ? (this.v = v, this) : new __await(v); }
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __asyncGenerator = (this && this.__asyncGenerator) || function (thisArg, _arguments, generator) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var g = generator.apply(thisArg, _arguments || []), i, q = [];
    return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
    function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
    function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
    function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
    function fulfill(value) { resume("next", value); }
    function reject(value) { resume("throw", value); }
    function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodeExplainerService = exports.ExplainOptions = void 0;
const memory_service_1 = __importDefault(require("./services/memory-service"));
const vector_terms_service_1 = __importDefault(require("./services/vector-terms-service"));
const completion_service_1 = require("./services/completion-service");
const interaction_history_1 = __importDefault(require("./interaction-history"));
const project_info_service_1 = __importDefault(require("./services/project-info-service"));
const code_selection_service_1 = __importDefault(require("./services/code-selection-service"));
const agent_1 = require("./agent");
const agent_selection_service_1 = __importDefault(require("./services/agent-selection-service"));
const lookup_context_service_1 = __importDefault(require("./services/lookup-context-service"));
const apply_context_service_1 = __importDefault(require("./services/apply-context-service"));
const classification_service_1 = __importDefault(require("./services/classification-service"));
const console_1 = require("console");
class ExplainOptions {
    constructor() {
        var _a, _b;
        this.modelName = (_a = process.env.APPMAP_NAVIE_MODEL) !== null && _a !== void 0 ? _a : 'gpt-4-0125-preview';
        this.tokenLimit = Number((_b = process.env.APPMAP_NAVIE_TOKEN_LIMIT) !== null && _b !== void 0 ? _b : 8000);
        this.temperature = 0.4;
        this.responseTokens = 1000;
    }
}
exports.ExplainOptions = ExplainOptions;
class CodeExplainerService {
    constructor(interactionHistory, completionService, classifierService, agentSelectionService, codeSelectionService, projectInfoService, memoryService) {
        this.interactionHistory = interactionHistory;
        this.completionService = completionService;
        this.classifierService = classifierService;
        this.agentSelectionService = agentSelectionService;
        this.codeSelectionService = codeSelectionService;
        this.projectInfoService = projectInfoService;
        this.memoryService = memoryService;
    }
    execute(clientRequest, options, chatHistory) {
        return __asyncGenerator(this, arguments, function* execute_1() {
            var _a, e_1, _b, _c;
            const { question: baseQuestion, codeSelection } = clientRequest;
            const classificationRequest = this.classifierService.classifyQuestion(baseQuestion);
            const projectInfoResponse = yield __await(this.projectInfoService.lookupProjectInfo());
            const projectInfo = Array.isArray(projectInfoResponse)
                ? projectInfoResponse
                : [projectInfoResponse];
            const { question, agent: mode } = this.agentSelectionService.selectAgent(baseQuestion, options, projectInfo);
            const tokensAvailable = () => options.tokenLimit - options.responseTokens - this.interactionHistory.computeTokenSize();
            const aggregateQuestion = [
                ...(chatHistory || [])
                    .filter((message) => message.role === 'user')
                    .map((message) => message.content),
                question,
                codeSelection,
            ]
                .filter(Boolean)
                .join('\n\n');
            const agentOptions = new agent_1.AgentOptions(question, aggregateQuestion, (chatHistory === null || chatHistory === void 0 ? void 0 : chatHistory.map((message) => message.content)) || [], projectInfo, codeSelection);
            yield __await(mode.perform(agentOptions, tokensAvailable));
            if (codeSelection)
                this.codeSelectionService.addSystemPrompt();
            const hasChatHistory = chatHistory && chatHistory.length > 0;
            if (hasChatHistory) {
                yield __await(this.memoryService.predictSummary(chatHistory));
            }
            if (codeSelection)
                this.codeSelectionService.applyCodeSelection(codeSelection);
            mode.applyQuestionPrompt(question);
            {
                const classification = yield __await(classificationRequest);
                (0, console_1.warn)(`Classification: ${classification}`);
            }
            const response = this.completionService.complete();
            try {
                for (var _d = true, response_1 = __asyncValues(response), response_1_1; response_1_1 = yield __await(response_1.next()), _a = response_1_1.done, !_a;) {
                    _c = response_1_1.value;
                    _d = false;
                    try {
                        const token = _c;
                        yield yield __await(token);
                    }
                    finally {
                        _d = true;
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (!_d && !_a && (_b = response_1.return)) yield __await(_b.call(response_1));
                }
                finally { if (e_1) throw e_1.error; }
            }
        });
    }
}
exports.CodeExplainerService = CodeExplainerService;
function explain(clientRequest, contextProvider, projectInfoProvider, helpProvider, options, chatHistory) {
    const interactionHistory = new interaction_history_1.default();
    const completionService = new completion_service_1.OpenAICompletionService(interactionHistory, options.modelName, options.temperature);
    const classificationService = new classification_service_1.default(interactionHistory, options.modelName, options.temperature);
    const codeSelectionService = new code_selection_service_1.default(interactionHistory);
    const vectorTermsService = new vector_terms_service_1.default(interactionHistory, options.modelName, options.temperature);
    const lookupContextService = new lookup_context_service_1.default(interactionHistory, contextProvider, helpProvider);
    const applyContextService = new apply_context_service_1.default(interactionHistory);
    const agentSelectionService = new agent_selection_service_1.default(interactionHistory, helpProvider, vectorTermsService, lookupContextService, applyContextService);
    const projectInfoService = new project_info_service_1.default(interactionHistory, projectInfoProvider);
    const memoryService = new memory_service_1.default(interactionHistory, options.modelName, options.temperature);
    const codeExplainerService = new CodeExplainerService(interactionHistory, completionService, classificationService, agentSelectionService, codeSelectionService, projectInfoService, memoryService);
    return {
        on: (event, listener) => {
            interactionHistory.on(event, listener);
        },
        execute() {
            return codeExplainerService.execute(clientRequest, options, chatHistory);
        },
    };
}
exports.default = explain;
