// @appland/navie/dist/services/agent-selection-service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const agent_1 = require("../agent");
const help_agent_1 = require("../agents/help-agent");
const generate_agent_1 = require("../agents/generate-agent");
const explain_agent_1 = __importDefault(require("../agents/explain-agent"));
const MODE_PREFIXES = {
    '@explain ': agent_1.AgentMode.Explain,
    '@generate ': agent_1.AgentMode.Generate,
    '@help ': agent_1.AgentMode.Help,
};
class AgentSelectionService {
    constructor(history, helpProvider, vectorTermsService, lookupContextService, applyContextService) {
        this.history = history;
        this.helpProvider = helpProvider;
        this.vectorTermsService = vectorTermsService;
        this.lookupContextService = lookupContextService;
        this.applyContextService = applyContextService;
    }
    selectAgent(question, options, projectInfo) {
        let modifiedQuestion = question;
        const hasAppMaps = projectInfo.some((info) => info.appmapStats.numAppMaps > 0);
        const helpAgent = () => new help_agent_1.HelpAgent(this.history, this.helpProvider, this.vectorTermsService);
        const generateAgent = () => new generate_agent_1.GenerateAgent(this.history, this.vectorTermsService, this.lookupContextService, this.applyContextService);
        const explainAgent = () => new explain_agent_1.default(this.history, this.vectorTermsService, this.lookupContextService, this.applyContextService);
        const buildAgent = {
            [agent_1.AgentMode.Help]: helpAgent,
            [agent_1.AgentMode.Generate]: generateAgent,
            [agent_1.AgentMode.Explain]: explainAgent,
        };
        const optionMode = () => {
            if (options.agentMode) {
                this.history.log(`[mode-selection] Activating agent due to explicit option: ${options.agentMode}`);
                const agent = buildAgent[options.agentMode]();
                return { question, agent };
            }
        };
        const questionPrefixMode = () => {
            for (const [prefix, mode] of Object.entries(MODE_PREFIXES)) {
                if (question.startsWith(prefix)) {
                    modifiedQuestion = question.slice(prefix.length);
                    this.history.log(`[mode-selection] Activating agent due to question prefix: ${mode}`);
                    const agent = buildAgent[mode]();
                    return { question: modifiedQuestion, agent };
                }
            }
        };
        const noAppMapsMode = () => {
            if (!hasAppMaps) {
                this.history.log(`[mode-selection] Activating Help mode because no AppMaps were detected.`);
                return { question, agent: helpAgent() };
            }
        };
        const defaultMode = () => {
            this.history.log(`[mode-selection] Using default mode: ${agent_1.AgentMode.Explain}`);
            return { question, agent: explainAgent() };
        };
        return optionMode() || questionPrefixMode() || noAppMapsMode() || defaultMode();
    }
}
exports.default = AgentSelectionService;
